src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/MessageBody.cs

```diff
@@ -217,6 +217,58 @@ private void ConsumedBytes(int count)
             OnConsumedBytes(count);
         }
 
+        internal static unsafe void ParseConnection(string connection, out bool isKeepAlive, out bool isUpgrade)
+        {
+            isKeepAlive = false;
+            isUpgrade = false;
+
+            fixed (char* ptr = connection)
+            {
+                var ch = ptr;
+                var end = ch + connection.Length;
+
+                while (ch < end)
+                {
+                    if ((*ch | 0x20) == 'k' && (end - ch) >= 9)
+                    {
+                        if ((*++ch | 0x20) == 'e' &&
```

Comment:
Could we cast to a ulongs and uints to compare 4 or 2 characters at a time? Can we use Benchmark.NET to determine comparing larger types is faster than the the character-by-character version, and that pointers in general are faster than `foreach`ing the string.

When I wrote similar code for `ValidateHeaderCharacters` I found foreach was faster than fixing the string and comparing using char pointers. You can see [the numbers I got here](https://github.com/aspnet/KestrelHttpServer/pull/804#issuecomment-217526770). The foreach version was `TestValidateHeaderValues1` and the `char*` version was `TestValidateHeaderValues2`